### ISSUE TITLE
Filter correctly implemented in .list()?

### ISSUE THREAD

**Issue by Im-Ev** [theophilusx/ssh2-sftp-client#468]
Hi,
Am I doing something wrong or is the filter function on sftp.list incorrect implemented?
On master, file index.js, line 498 shouldn't it be:
`resolve(newList.filter(filter);`

instead of:
`resolve(newList.filter((item) => filter(item)));`

**Comment by theophilusx** [comment]
The two lines are functionally equivalent. e.g.


const f = (i) => {
  return i.startsWith("Bo");
}

let data = ['Bob', 'Boris', 'Bill', 'Ben', 'Bo'];

let r1 = data.filter(f);
let r2 = data.filter((item) => f(item));

console.log('r1', r1);
console.log('r2', r2);

Gives the result

r1 [ 'Bob', 'Boris', 'Bo' ]
r2 [ 'Bob', 'Boris', 'Bo' ]

This would indicate your doing something wrong in your filter
function. However, without seeing the function or the error message, I
have no clue.

**Comment by Im-Ev** [comment]
Hi, thanks for the answer.
I'm calling the list function like this:
`sftp.list('/path/to/remote/dir', (i) => { return /foo*bar/.test(i) });`

Shouldn't this be valid? Because I don't get any output from that.

**Comment by theophilusx** [comment]
Im-Ev ***@***.***> writes:

> Hi, thanks for the answer.
> I'm calling the list function like this:
> sftp.list('/path/to/remote/dir', (i) => { return /foo*bar/.test(i) });
>
> Shouldn't this be valid? Because I don't get any output from that.
>

What is /foobar/.test(i) ???

That doesn't look correct. What you have is equivalent to

function blah(i) {
  return /foobar/.test(i);
}

Is that a valid function? I doesn't look correct to me.

Note that list.filter((x) => someFunc(x)) is the same (roughly) as 
list.filter(someFunc);

The point is, 'someFunc' needs to be a working function which accepts
one argument. In this specific example, the argument is one item from
the list() output array. The items of this array are objects which
describe each entry i.e. filename, mode, owner, gorup etc.

So what you need to do define a filter function which accepts one
argument, looks at that argument in some way (i.e. looks at the name
property of the object) and either returns true or false. If it returns
true, keep that item, false and it is filtered out.

You should be able to run that function at the repl and just pass it in
a list object (see the docs for the properties in the list object).

Suggest you create a test array of objects and experiment with the
filter method until you get something which works.

**Comment by Im-Ev** [comment]
Yes, `/foobar/.test(i)` is a valid function. It's a RegEx statement see: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/test

Like the array.test function it also returns true or false if the given test function matches.

My implemented code looks like this (including debugging logs):
`sftp.list(ftpDir, (el: any) => {
   console.log(`Current element: ${el.name}`);
   if (/foobar/.test(el.name)) {
     console.log(`Found a match`);
     return true;
   }
   return false;
})`

The logged output is:
`Current element: hello
Current element: foobar
Found a match`

The output what I get from sftp.list:
`[{"type":"d","name":"hello","modifyTime":1676542847000,"accessTime":1676384110000,"rights:{"user":"rwx","group":"rwx","other":"rwx"},"longname":"drwxrwxrwx   1 10027    10003           0 Feb 16 11:20 hello"},{"type":"d","name":"foobar","modifyTime":1677041567000,"accessTime":1676985269000,"rights":{"user":"rwx","group":"rwx","other":"rwx"},"longname":"drwxrwxrwx   1 10027    10003           0 Feb 22 05:52 foobar"}]`

The output which I would expect:
`{"type":"d","name":"foobar","modifyTime":1677041567000,"accessTime":1676985269000,"rights":{"user":"rwx","group":"rwx","other":"rwx"},"longname":"drwxrwxrwx   1 10027    10003           0 Feb 22 05:52 foobar"}`

For my understanding this should work but I don't know why it doesn't.

**Comment by theophilusx** [comment]
FYI I cannot reproduce the issue you are seeing. Below are the test cases I'm using which are using a filter very similar to what you were trying to do. They work without error.  This is tested against current master and the tests are in the test directory of the repository if you want to look closer. 

```js
describe('list tests with filters', function () {
  let sftp;

  before('list() test setup hook', async function () {
    sftp = await getConnection();
    await sftp.put(Buffer.from('A foo file'), `${config.sftpUrl}/foo1.txt`, {
      encoding: 'utf8',
    });
    await sftp.put(Buffer.from('A scond foo file'), `${config.sftpUrl}/foo2.txt`, {
      encoding: 'utf8',
    });
    await sftp.put(Buffer.from('A bar file'), `${config.sftpUrl}/bar1.txt`, {
      encoding: 'utf8',
    });
    await sftp.put(Buffer.from('A second bar file'), `${config.sftpUrl}/bar2.txt`, {
      encoding: 'utf8',
    });
    await sftp.put(Buffer.from('A baz file'), `${config.sftpUrl}/baz1.txt`, {
      encoding: 'utf8',
    });
    await sftp.put(Buffer.from('A second bar file'), `${config.sftpUrl}/baz2.txt`, {
      encoding: 'utf8',
    });
    return true;
  });

  after('list() test cleanup hook', async function () {
    await sftp.delete(`${config.sftpUrl}/foo1.txt`);
    await sftp.delete(`${config.sftpUrl}/foo2.txt`);
    await sftp.delete(`${config.sftpUrl}/bar1.txt`);
    await sftp.delete(`${config.sftpUrl}/bar2.txt`);
    await sftp.delete(`${config.sftpUrl}/baz1.txt`);
    await sftp.delete(`${config.sftpUrl}/baz2.txt`);
    await sftp.end();
    return true;
  });

  it('basic listing', async function () {
    let files = await sftp.list(config.sftpUrl);
    return expect(files).containSubset([
      { name: 'foo1.txt' },
      { name: 'foo2.txt' },
      { name: 'bar1.txt' },
      { name: 'bar2.txt' },
      { name: 'baz1.txt' },
      { name: 'baz2.txt' },
    ]);
  });

  it('listing of just foo files', async function () {
    let flt = (item) => {
      return /foo.\.txt/.test(item.name);
    };
    let files = await sftp.list(config.sftpUrl, flt);
    return expect(files).containSubset([{ name: 'foo1.txt' }, { name: 'foo2.txt' }]);
  });

  it('listing only version 1 files', async function () {
    let flt = (item) => {
      return /.*1\.txt/.test(item.name);
    };
    let files = await sftp.list(config.sftpUrl, flt);
    return expect(files).containSubset([
      { name: 'foo1.txt' },
      { name: 'bar1.txt' },
      { name: 'baz1.txt' },
    ]);
  });
});
```
